recipe/build.sh

```diff
@@ -57,9 +57,17 @@ make
 # change permissions again after building
 chmod -R o-w "${SRC_DIR}"
 
+# 2017/6/17:
 # Seems we hit:
 # lib/perlbug .................................................... # Failed test 21 - [perl \#128020] long body lines are wrapped: maxlen 1157 at ../lib/perlbug.t line 154
 # FAILED at test 21
 # https://rt.perl.org/Public/Bug/Display.html?id=128020
+
+# 2023/12/18:
+# t/re/substT ...................................................... FAILED--expected 278 tests, saw 161
```

Comment:
Could you add a brief justfication why the test failures/omissions aren't a problem?

Reply:
Looks like **v5.35.5** will require to rebuild [all downstream packages](https://metayaml-conda.fly.dev/metayaml/reverse_dependencies?name=perl) because:
`Perl 5.35.5 is not binary compatible with earlier versions of Perl.  `
See https://metacpan.org/release/LEONT/perl-5.35.5/view/INSTALL#Coexistence-with-earlier-versions-of-perl-5 and https://metacpan.org/release/LEONT/perl-5.35.5/view/INSTALL#Upgrading-from-5.35.4-or-earlier
I wouldn't do that without a plan now as it will be challenging.

Reply:
We never did `make check` because it was always disabled. I enabled it to see which errors will be for this release. 
Perl's test failures are quite enigmatic to me. I appreciate any help to debug them

Reply:
I ran `make test_harness` for the verbose output:

```
Test Summary Report
-------------------
re/substT.t                                                        (Wstat: 0 Tests: 161 Failed: 0)
  Parse errors: Bad plan.  You planned 278 tests but ran 161.
op/utftaint.t                                                      (Wstat: 512 Tests: 86 Failed: 0)
  Non-zero exit status: 2
  Parse errors: Bad plan.  You planned 89 tests but ran 86.
porting/libperl.t                                                  (Wstat: 7424 Tests: 0 Failed: 0)
  Non-zero exit status: 29
  Parse errors: No plan found in TAP output
../ext/Devel-Peek/t/Peek.t                                         (Wstat: 512 Tests: 107 Failed: 0)
  Non-zero exit status: 2
  Parse errors: No plan found in TAP output
../ext/DynaLoader/t/DynaLoader.t                                   (Wstat: 0 Tests: 44 Failed: 1)
  Failed test:  20
../lib/overload.t                                                  (Wstat: 512 Tests: 5011 Failed: 0)
  Non-zero exit status: 2
  Parse errors: Bad plan.  You planned 5363 tests but ran 5011.
../lib/perlbug.t                                                   (Wstat: 0 Tests: 25 Failed: 1)
  Failed test:  21
Files=2672, Tests=1196897, 347 wallclock secs (37.40 usr  9.27 sys + 329.04 cusr 77.29 csys = 453.00 CPU)
Result: FAIL
make: *** [GNUmakefile:849: test_harness] Error 2
```

Reply:
Also in detail:
```
# Failed test 21 - [perl \#128020] long body lines are wrapped: maxlen 1521 at ../lib/perlbug.t line 154
../lib/perlbug.t ..................................................... 
```

```
# Failed test 20 - array should contain one result or more: libc => () at t/DynaLoader.t line 127
#      got "0"
# expected >= "1"
../ext/DynaLoader/t/DynaLoader.t ..................................... 
```

```
Insecure directory in $ENV{PATH} while running with -T switch at ../../t/test.pl line 810.
# Tests were run but no plan was declared and done_testing() was not seen.
# Looks like your test exited with 2 just after 107.
../ext/Devel-Peek/t/Peek.t ........................................... 

```

```
Insecure directory in $ENV{PATH} while running with -T switch at ./test.pl line 810, <$iter_fh> line 18.
# Looks like you planned 5363 tests but ran 5011.
../lib/overload.t .................................................... 
```

Seems like `Insecure directory in $ENV{PATH}` is a common issue and something is wrong with libc/dynamic loader

Reply:
Fair enough about the rebuild issue. Might be worth upgrading to latest at some point

Re the testing. The insecure directory failures could be some platform/environment thing to do with us. The failed tests 20 and 21 look like they could do with a bit more investigation. Maybe the "bad plan" things are an easy fix or relatively easy to understand (looks like test configuration).

Personally I put quite a lot of time in to test medium-to-high impact compiled libraries and reason about test failures. Maybe too much time. It would be helpful and useful to have some opinions from some other engineers about this, maybe @JeanChristopheMorinPerso, @rafaelmartins or maybe even @chenghlee if he's got the time

Reply:
i.e. at what point to stop investigating and just note the failures, even if we haven't got to the bottom of it

Reply:
or @markan

Reply:
perl upgrades are kinda complicated, because you likely need to rebuild revdeps, specially libraries, even if upstream says that there's binary compatibility, because the libraries are installed to a folder that includes the full version number. Rebuilding everything is usually the easier thing to do, even for minor releases.

Reply:
about the tests, the "correct" way to do that is to skip the tests that are supposed to fail, but still run the testsuite every time. some test toolkits support skipping a test from command line, but usually what is done is to patch out any tests that are supposed to fail, and in the patch add information on why those tests are not passing. not running the whole testsuite is bad because you could be missing unexpected failures.

Reply:
the message about insecure directory does not seem to be a loader problem, it is a warning because apparently conda is including some world writtable folder (probably a tmpdir) to the PATH. I don't think that it is the final error.

Reply:
Yeah, I'm thinking primarily about how to determine which test failures are ok (supposed to fail) and which aren't

Reply:
but yeah good point, `make test` shouldn't just be commented out

Reply:
you could check if someone else also skips that test, like debian, gentoo, etc. but usually you'd need to do the analysis and find that out yourself. you need to find a reason why that test is ok to fail in our env. for example, if that insecure folder thing was an error, if I know that conda adds a folder like that to the PATH and I have no way to avoid that, I'm ok skipping such test. there's not much shortcut for this

Reply:
>you need to find a reason why that test is ok to fail in our env

so in other words, we should treat as compulsory:
a) running the upstream tests (with some reasonable coverage)
b) explicitly skipping failing tests
c) reasoning about why all test failures are ok

Reply:
(is that the approach in yum/rpm?)

Reply:
yes, that's how pretty much every mainstream linux distro handles this.

Reply:
@skupr-anaconda we could pair on this tomorrow evening your time if you would like

Reply:
@danpetry Sure, it will be great